zcash_keys/src/keys.rs

```diff
@@ -791,43 +833,262 @@ impl UnifiedFullViewingKey {
+        );
+        #[cfg(feature = "sapling")]
+        let items = items.chain(
+            self.sapling
+                .as_ref()
+                .map(|divk| divk.to_bytes())
+                .map(unified::Ivk::Sapling),
+        );
+        #[cfg(feature = "transparent-inputs")]
+        let items = items.chain(
+            self.transparent
+                .as_ref()
+                .map(|tivk| tivk.serialize().try_into().unwrap())
+                .map(unified::Ivk::P2pkh),
+        );
+
+        unified::Uivk::try_from_items(items.collect())
+            .expect("UnifiedIncomingViewingKey should only be constructed safely.")
+    }
+
+    /// Returns the Transparent external IVK, if present.
+    #[cfg(feature = "transparent-inputs")]
+    fn transparent(&self) -> &Option<zcash_primitives::legacy::keys::ExternalIvk> {
+        &self.transparent
+    }
+
+    /// Returns the Sapling IVK, if present.
+    #[cfg(feature = "sapling")]
+    fn sapling(&self) -> &Option<::sapling::zip32::IncomingViewingKey> {
+        &self.sapling
+    }
+
+    /// Returns the Orchard IVK, if present.
+    #[cfg(feature = "orchard")]
+    fn orchard(&self) -> &Option<orchard::keys::IncomingViewingKey> {
+        &self.orchard
+    }
+
+    /// Attempts to derive the Unified Address for the given diversifier index and
+    /// receiver types.
+    ///
+    /// Returns `None` if the specified index does not produce a valid diversifier.
+    pub fn address(
+        &self,
+        j: DiversifierIndex,
+        request: UnifiedAddressRequest,
     ) -> Result<UnifiedAddress, AddressGenerationError> {
         #[cfg(feature = "orchard")]
         let mut orchard = None;
-        if _request.has_orchard {
+        if request.has_orchard {
             #[cfg(not(feature = "orchard"))]
             return Err(AddressGenerationError::ReceiverTypeNotSupported(
                 Typecode::Orchard,
             ));
 
             #[cfg(feature = "orchard")]
-            if let Some(ofvk) = &self.orchard {
-                let orchard_j = orchard::keys::DiversifierIndex::from(*_j.as_bytes());
-                orchard = Some(ofvk.address_at(orchard_j, Scope::External))
+            if let Some(oivk) = &self.orchard {
+                let orchard_j = orchard::keys::DiversifierIndex::from(*j.as_bytes());
+                orchard = Some(oivk.address_at(orchard_j))
             } else {
                 return Err(AddressGenerationError::KeyNotAvailable(Typecode::Orchard));
             }
         }
 
         #[cfg(feature = "sapling")]
         let mut sapling = None;
-        if _request.has_sapling {
+        if request.has_sapling {
             #[cfg(not(feature = "sapling"))]
             return Err(AddressGenerationError::ReceiverTypeNotSupported(
                 Typecode::Sapling,
             ));
 
             #[cfg(feature = "sapling")]
-            if let Some(extfvk) = &self.sapling {
+            if let Some(extivk) = &self.sapling {
```

Comment:
nonblocking nit: the `ext` here is unnecessary (and was actually kind of incorrect previously, as it should have been named `divk`).